experiments/pop_test.go

```diff
@@ -0,0 +1,255 @@
+		},
+	})
+
+	b, err := builder.Build()
+	require.NoError(t, err)
+
+	t.Logf("server generated biscuit:\n%s", b.String())
+
+	s, err := b.Serialize()
+	require.NoError(t, err)
+	return s, serverKey.Public()
+}
+
+func clientSign(t *testing.T, rootPubkey sig.PublicKey, pubkey ed25519.PublicKey, privkey ed25519.PrivateKey, b []byte) []byte {
+	token, err := biscuit.Unmarshal(b)
+	require.NoError(t, err)
+
+	verifier, err := token.Verify(rootPubkey)
+	require.NoError(t, err)
+
+	t.Logf("clientSign world:\n%s", verifier.PrintWorld())
+
+	// This query returns to_sign(dataID, alg, data) facts which require a signature with a private key matching pubkey.
+	// in this example: [to_sign(0, "ed25519", "hex:7369676e2074686973")]
+	toSign, err := verifier.Query(biscuit.Rule{
+		Head: biscuit.Predicate{Name: "to_sign", IDs: []biscuit.Atom{biscuit.Variable(0), biscuit.Variable(1), biscuit.Variable(2)}},
+		Body: []biscuit.Predicate{
+			{Name: "should_sign", IDs: []biscuit.Atom{biscuit.SymbolAuthority, biscuit.Variable(0), biscuit.Variable(1), biscuit.Bytes(pubkey)}},
+			{Name: "data", IDs: []biscuit.Atom{biscuit.SymbolAuthority, biscuit.Variable(0), biscuit.Variable(2)}},
+		},
+	})
+	require.NoError(t, err)
+	t.Logf("toSign:\n%#v", toSign)
+	require.Equal(t, 1, len(toSign))
+
+	// Extract data from the fact
+	dataID, ok := toSign[0].IDs[0].(biscuit.Integer)
+	require.True(t, ok)
+	alg, ok := toSign[0].IDs[1].(biscuit.Symbol)
+	require.True(t, ok)
+	data, ok := toSign[0].IDs[2].(biscuit.Bytes)
+	require.True(t, ok)
+
+	// We have a "to_sign" fact, so we check if the token doesn't already hold a signature:
+	alreadySigned, err := verifier.Query(biscuit.Rule{
+		Head: biscuit.Predicate{Name: "already_signed", IDs: []biscuit.Atom{biscuit.Variable(0)}},
+		Body: []biscuit.Predicate{
+			{Name: "signature", IDs: []biscuit.Atom{dataID, alg, biscuit.Variable(0)}},
+		},
+	})
+	require.NoError(t, err)
+	t.Logf("already signed:\n%#v", alreadySigned)
+	require.Equal(t, 0, len(alreadySigned))
+
+	signerTimestamp := time.Now()
+	signerNonce := make([]byte, 32)
+	_, err = rand.Read(signerNonce)
+	require.NoError(t, err)
+
+	tokenHash, err := token.Sha256Sum(token.BlockCount())
+	require.NoError(t, err)
+
+	dataToSign := append(data, tokenHash...)
+	dataToSign = append(dataToSign, signerNonce...)
+	dataToSign = append(dataToSign, []byte(signerTimestamp.Format(time.RFC3339))...)
+
+	// Sign the data
+	var signedData biscuit.Bytes
+	switch alg {
+	case "ed25519":
+		signedData = ed25519.Sign(privkey, dataToSign)
+	default:
+		t.Fatalf("unsupported alg: %s", alg)
+	}
+
+	// Add the signature to the token
+	builder := token.CreateBlock()
+
+	err = builder.AddFact(biscuit.Fact{Predicate: biscuit.Predicate{
+		Name: "signature",
+		IDs:  []biscuit.Atom{dataID, alg /*biscuit.Bytes(pubkey),*/, signedData},
+	}})
+	require.NoError(t, err)
+	// not sure we need to repeat the pubkey here,
+	// since the one from the authority fact can be used for the verification.
```

Comment:
I think the reason for this is that we could have multiple signatures required across the same data in a single biscuit.